### PR TITLE
Add polkadot testnet network (Westend)

### LINKED ISSUE #4248: Add Westend polkadot testnet network
Add westend network support to freely test polkadot.

**Denom:**

**Westie (WND)**
ss58 format: 42
Decimals: 12

**Public node:**

wss://westend-rpc.polkadot.io

**Faucet:**

The current testnet is called Westend and you can obtain its native tokens by posting !drip <WESTEND_ADDRESS> in the Riot chatroom #westend_faucet:matrix.org.

**Telemetry:**

https://telemetry.polkadot.io/#list/Westend

**Reference:**

https://polkadot.network/westend-introducing-a-new-testnet-for-polkadot-and-kusama/

### PR DESCRIPTION
**Pull request by mariopino**
Closes #4248 

Need to test send/stake/unstaking/all_the_things but fauce seems not operative, still waiting for my fresh WND tokens to test.

**Faucet**

The current testnet is called Westend and you can obtain its native tokens by posting !drip <WESTEND_ADDRESS> in the Riot chatroom #westend_faucet:matrix.org.

### CODE REVIEW

**Comment by faboweb** [comment]
🚀 Deployed on https://5ee1e88f1845d392a3697505--lunieio.netlify.app

**Comment by mariopino** [comment]
Tested Send transaction and it fails! :-/

```
2020-06-11 17:40:27        RPC-CORE: submitExtrinsic(extrinsic: Extrinsic): Hash:: 1010: Invalid Transaction: Inability to pay some fees (e.g. account balance too low)
```

I have enough balance (0.3 WND) and was trying to send 0,000001 sooo time to debug!

File: api/data/networks.js [hunk 1]
```diff
@@ -222,6 +222,32 @@ module.exports = [
     slug: 'polkadot',
     lockUpPeriod: '28 days'
   },
+  {
+    id: 'westend',
```
Comment (faboweb) [inline api/data/networks.js:226]: polkadot-testnet?

**Comment by mariopino** [comment]
Found type not found errors:

```
Unknown types found, no types for Status
```

Upgrading polkadot-api to v1.19.0-beta.3 fixes them!

File: api/lib/controller/transaction/index.js [hunk 2]
```diff
@@ -7,21 +7,54 @@ const networks = require('../../../data/networks')
 
 global.fetch = require('node-fetch')
 
-let api
-async function initPolkadotAPI() {
+let kusamaAPI, polkadotAPI, polkadotTestnetAPI
```
Comment (faboweb) [inline api/lib/controller/transaction/index.js:10]: 🤮 
there is no data structure that scales? like an object keyed by the networkId?

File: app/src/components/common/TmSessionSignIn.vue [hunk 3]
```diff
@@ -72,11 +72,17 @@ import TmField from "common/TmField"
 import TmFormMsg from "common/TmFormMsg"
 import TmFormStruct from "common/TmFormStruct"
 import SessionFrame from "common/SessionFrame"
+import { checkAddress } from "@polkadot/util-crypto"
+
 const isPolkadotAddress = (address) => {
```
Comment (faboweb) [inline app/src/components/common/TmSessionSignIn.vue:77]: is this still used?

File: app/src/signing/networkMessages/polkadot-testnet-transactions.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (faboweb) [inline app/src/signing/networkMessages/polkadot-testnet-transactions.js:1]: this file is duplicating a lot of code. can we refactor this?

File: app/src/signing/networkMessages/polkadot-testnet.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
+import { getSignMessage, getAPI } from "./polkadot-testnet-transactions"
```
Comment (faboweb) [inline app/src/signing/networkMessages/polkadot-testnet.js:1]: same here. can we refactor this?

File: app/src/signing/networkMessages/polkadot.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
+import { getSignMessage, getAPI } from "./polkadot-transactions"
```
Comment (faboweb) [inline app/src/signing/networkMessages/polkadot.js:1]: same

File: api/lib/controller/transaction/index.js (hunk 2 above)
Comment (Bitcoinera) [reply api/lib/controller/transaction/index.js:10]: :sweat_smile:  the vomit emoji

**Comment by Bitcoinera** [comment]
Still not working

File: app/src/components/common/TmSessionSignIn.vue (hunk 3 above)
Comment (faboweb) [reply app/src/components/common/TmSessionSignIn.vue:77]: ok let's leave in for now

File: app/src/gql/index.js [hunk 7]
```diff
@@ -136,6 +136,7 @@ export const NetworksAll = gql`
         viewDenom
         chainToViewConversionFactor
       }
+      rpc_url
```
Comment (faboweb) [inline app/src/gql/index.js:139]: hmm doesn't work as the rpc usually is whitelisted to the BE only. let's create a dictionary in the FE for those?

File: app/src/signing/networkMessages/polkadot-transactions.js [hunk 8]
```diff
@@ -34,6 +44,7 @@ export async function getAPI(endpoint) {
 // }
 
 export async function multiMessage(transactions) {
+  const polkadotNetworks = await getPolkadotNetworks()
```
Comment (faboweb) [inline app/src/signing/networkMessages/polkadot-transactions.js:47]: this would be called every time. why not just get the normal networks list?

**Comment by faboweb** [comment]
Consider that the extension will also need to provide the api url now

File: app/src/gql/index.js (hunk 7 above)
Comment (Bitcoinera) [reply app/src/gql/index.js:139]: Ah no sorry, I didn't update, it is working now

File: app/src/signing/networkMessages/polkadot-transactions.js (hunk 8 above)
Comment (Bitcoinera) [reply app/src/signing/networkMessages/polkadot-transactions.js:47]: from where? There is no normal networks list, only in the store's state... But not sure how that would work out here

Comment (faboweb) [reply app/src/signing/networkMessages/polkadot-transactions.js:47]: we only need the rpc_urls right? let's go with just a dictionary for the rpc_urls

File: app/src/gql/index.js (hunk 7 above)
Comment (faboweb) [reply app/src/gql/index.js:139]: no I mean we need to use the public polkadot endpoints in the FE

File: app/src/signing/networkMessages/polkadot-transactions.js (hunk 8 above)
Comment (Bitcoinera) [reply app/src/signing/networkMessages/polkadot-transactions.js:47]: but we need to update them no? What happens if they suddenly change? Then we need to remember updating them here too

File: app/src/gql/index.js (hunk 7 above)
Comment (Bitcoinera) [reply app/src/gql/index.js:139]: See my comment below: https://github.com/luniehq/lunie/pull/4266#discussion_r444560139

File: app/src/signing/networkMessages/polkadot-transactions.js (hunk 8 above)
Comment (faboweb) [reply app/src/signing/networkMessages/polkadot-transactions.js:47]: True! We could pass public rpc urls independent from our internal rpc urls. But send both from the Api

Comment (Bitcoinera) [reply app/src/signing/networkMessages/polkadot-transactions.js:47]: You mean creating a query for rpcUrls specifically? Isn't is simpler to just stick to the current networks query?

Comment (faboweb) [reply app/src/signing/networkMessages/polkadot-transactions.js:47]: No I mean returning the public endpoints from the API for the networks query. We could define a property in the networks "public_rpc_endpoint" and only return that.

File: api/lib/schema.js [hunk 9]
```diff
@@ -133,7 +133,7 @@ const typeDefs = gql`
     id: String
     title: String
     chain_id: String
-    rpc_url: String
+    public_rpc_url: String
```
Comment (Bitcoinera) [inline api/lib/schema.js:136]: This is wrong right. Why do you delete `rpc_url` here. Now you get the following error:

image

File: api/tests/network-configs.test.js [hunk 10]
```diff
@@ -11,7 +11,7 @@ const schema = Joi.object({
   id: Joi.string().lowercase(),
   title: Joi.string(),
   chain_id: Joi.string(),
-  rpc_url: Joi.string().uri(),
+  public_rpc_url: Joi.string().uri(),
```
Comment (Bitcoinera) [inline api/tests/network-configs.test.js]: Same than above

File: extension/src/popup/gql/index.js [hunk 11]
```diff
@@ -15,7 +15,7 @@ export const Networks = gql`
       chain_id
       testnet
       title
-      rpc_url
+      public_rpc_url
```
Comment (Bitcoinera) [inline extension/src/popup/gql/index.js:18]: To be honest, I still don't quite understand the why for having `public_rpc_url`, but it's OK

File: app/src/signing/networkMessages/polkadot-transactions.js (hunk 8 above)
Comment (Bitcoinera) [reply app/src/signing/networkMessages/polkadot-transactions.js]: But why not just return the rpc_urls? I think I'm missing something here. Will keep looking at the commits

File: api/lib/schema.js (hunk 9 above)
Comment (faboweb) [reply api/lib/schema.js:136]: let's revert I changed it actually to return the public url

File: api/tests/network-configs.test.js (hunk 10 above)
Comment (faboweb) [reply api/tests/network-configs.test.js]: also revert

Comment (faboweb) [reply api/tests/network-configs.test.js]: ah no, we want both here

File: extension/src/popup/gql/index.js (hunk 11 above)
Comment (faboweb) [reply extension/src/popup/gql/index.js:18]: the internal rpc url is blocked for users.

Comment (Bitcoinera) [reply extension/src/popup/gql/index.js:18]: oh cool

File: api/tests/network-configs.test.js (hunk 10 above)
Comment (Bitcoinera) [reply api/tests/network-configs.test.js]: OK, on it

File: api/lib/resolvers.js [hunk 12]
```diff
@@ -340,16 +340,10 @@ const resolvers = {
...
-            // access the docker container from the outside via its port)
-            return {
-              ...network,
-              api_url: 'http://localhost:9071'
-            }
-          }
-          return network
+          return Object.assign({}, network, {
+            rpc_url: network.public_rpc_url,
+            public_rpc_url: undefined
```
Comment (Bitcoinera) [inline api/lib/resolvers.js:345]: Why this? This is what is making get `null` in the FE for `public_rpc_url`. What was your idea here?

Comment (faboweb) [reply api/lib/resolvers.js:345]: the idea is to only output `rpc_url` which is the public url. the frontend doesn't need to know about the private url

Comment (faboweb) [reply api/lib/resolvers.js:345]: need to adjust in polkadot-transactions line 32

Comment (Bitcoinera) [reply api/lib/resolvers.js:345]: ok I see

Comment (Bitcoinera) [reply api/lib/resolvers.js:345]: Shouldn't we do the same for the single network query?

**Pull request merged**